### PR TITLE
Add the argument bounds to the .scope and .plotTree methods to define their bounds; add the argument position to the .meter method to define its position

### PR DESCRIPTION
**Pull request by prko**
## Purpose and Motivation
`.scope` has no argument to define its window location. Often it is good to see this window on the centre of the screen, but sometimes it is not so optical. 
This modification adds one more argument for this feature. Thus, the following code is available (I intentionally paste all code to present the necessity of this feature):
```
(
Window.closeAll;
s.waitForBoot {
	var sigs = {
		[
			SinOsc.ar(440),
			(SinOsc.ar(440 * (1..48)) * (1..48).reciprocal / 16).sum,
			Resonz.ar(Resonz.ar(WhiteNoise.ar, 440, 0.1, 5), 440, 0.2, 5),
			Resonz.ar(
				Resonz.ar(WhiteNoise.ar, 440 * (1..48), 0.2, 5),
				440 * (1..48), 0.2, (1..48).reciprocal
			).sum,
			WhiteNoise.ar
		] * Env.perc(0.01, 0.79, 0.2).ar(Done.freeSelf)
	};
	FreqScope();
	sigs.().size.do { |i|
		{ sigs.()[i] ! 2 }.scope(position: 605@0); // <- look at this!
		{ sigs.()[i] }.plot(0.04, bounds:Rect(230 * i, 355 + 175, 230, 150));
		{ sigs.()[i] }.plot(1, bounds:Rect(230 * i, 355, 230, 150));
		2.wait
	}
}
)
``` 
demo video:
https://www.dropbox.com/scl/fi/qj0337im08mqyzcj2k4fx/Screen-Recording-2024-03-06-at-14.03.44.mov?rlkey=k82z5kwoi2w0nj23sy56nfxht&dl=0

## Types of changes

- Documentation
- New feature

## To-do list

- [x] Code is tested
- [x] All tests are passing
- [x] Updated documentation
- [x] This PR is ready for review

### CODE REVIEW

**Comment by mtmccrea** [comment]
Typically a window widget would accept a `Rect` for it's position/size, as opposed to just a position `Point`.

**Comment by prko** [comment]
@mtmccrea 
Thanks! I fixed all the problems!

I also did the same for Server plotTree and Server meter. However, I am not sure if I should upload these changes here or make two other separate PRs.
https://www.dropbox.com/scl/fo/7jwlyoipy5zg4se2o7kmh/h?rlkey=eo7an9nz4wv1dn2jpze0ywsj8&dl=0

**Comment by prko** [comment]
@mtmccrea 
Thanks! I fixed all the problems!

I also did the same for Server plotTree and Server meter. However, I am not sure if I should upload these changes here or make two other separate PRs.
https://www.dropbox.com/scl/fi/6wzqzldir2kebqjx5jx20/SC-bounds-added-for-plotTree-scope-and-meter.mov?rlkey=8rpxmwrz1et6659e59zrl1fe9&dl=0

**Comment by prko** [comment]
The `bounds` argument has also been added to the `.plotTree` method.

The size of `s.meter` is automatically calculated by the number of input and output channels. So I think it would be better to give a point argument instead of bounds. However, I will wait for a response from other contributors for a few days. @mtmccrea What do you think?

**Review by telephon — CHANGES_REQUESTED** [review]
The idea is reasonable, please add the documentation.

**Comment by prko** [comment]
@telephon 
@mtmccrea 
I interpreted the comment and change request from @telephon as an approval of the position argument for the `.meter` method. So, I added this and updated the documentation. If I misunderstood @telephon 's change request, please let me know!

File: SCClassLibrary/Common/GUI/PlusGUI/Control/server-scope.sc
Comment (mtmccrea) [inline SCClassLibrary/Common/GUI/PlusGUI/Control/server-scope.sc:15]: Looks like you could avoid the duplicated code and extra logic caused by `if(bounds.isNil)` by letting `scopeWindow` be created as before, then if the `bounds` arg is provided, access the window enclosing `scopeWindow` with `scopeWindow.window.bounds_(bounds)`. (untested)

That way responsibility for creating the window stays in one place.

File: SCClassLibrary/Common/GUI/tools/Stethoscope.sc
Comment (mtmccrea) [inline SCClassLibrary/Common/GUI/tools/Stethoscope.sc:404]: If this change is approved, this behavior should be documented: it changes the window position in favor of keeping the window fully on-screen with specified size.

File: SCClassLibrary/Common/GUI/PlusGUI/Control/ServerPlusGUI.sc
Comment (mtmccrea) [inline SCClassLibrary/Common/GUI/PlusGUI/Control/ServerPlusGUI.sc:371]: spaces within curly brackets:
`{ defer { plotTreeWindow.close } }`

**Comment by mtmccrea** [comment]
Something to consider: what would the code look like to change the bounds of the GUI without the new argument? 
If it's just a matter of calling `s.scope.bounds_(Rect(1,2,3,4))`, `s.meter.bounds_(Rect(1,2,3,4))`, or similar (can't test atm) then it may not be worth adding?

File: SCClassLibrary/Common/GUI/PlusGUI/Control/ServerPlusGUI.sc
Comment (prko) [reply SCClassLibrary/Common/GUI/PlusGUI/Control/ServerPlusGUI.sc]: Thanks! Done!

**Comment by prko** [comment]
@mtmccrea 
> If it's just a matter of calling `s.scope.bounds_(Rect(1,2,3,4))`, `s.meter.bounds_(Rect(1,2,3,4))`, or similar (can't test atm) then it may not be worth adding?

Each of them returns the following error:
>ERROR: Message 'bounds_' not understood.

The full error message for  `s.scope.bounds_(Rect(200,200,400,400))`:
```
ERROR: Message 'bounds_' not understood.
Perhaps you misspelled 'bus_', or meant to call 'bounds_' on another receiver?
RECEIVER:
Instance of Stethoscope {    (000002D58E240A18, gc=A8, fmt=00, flg=00, set=05)
  instance variables [32]
    playSynthDef : nil
    makeGui : instance of Function (000002D58E7397C8, size=2, set=2)
    setCycle : instance of Function (000002D58E7394F8, size=2, set=2)
    setYZoom : instance of Function (000002D58E739488, size=2, set=2)
    setIndex : instance of Function (000002D58E7391B8, size=2, set=2)
    setNumChannels : instance of Function (000002D58E739148, size=2, set=2)
    setRate : instance of Function (000002D58E738E78, size=2, set=2)
    setStyle : instance of Function (000002D58E738E08, size=2, set=2)
    updateColors : instance of Function (000002D58E738B38, size=2, set=2)
    window : instance of Window (000002D58D8FCA68, size=5, set=3)
    view : instance of Window (000002D58D8FCA68, size=5, set=3)
    scopeView : instance of ScopeView (000002D58E8C1C18, size=41, set=6)
    cycleSlider : instance of Slider (000002D58E8C2098, size=39, set=6)
    yZoomSlider : instance of Slider (000002D58E8C2518, size=39, set=6)
    idxNumBox : instance of NumberBox (000002D58E8C2E18, size=46, set=6)
    chNumBox : instance of NumberBox (000002D58E8C3298, size=46, set=6)
    styleMenu : instance of PopUpMenu (000002D58E8C3718, size=35, set=6)
    rateMenu : instance of PopUpMenu (000002D58E8C2998, size=35, set=6)
    server : instance of Server (000002D58ADAB238, size=31, set=5)
    synth : instance of BusScopeSynth (000002D58D901D98, size=5, set=3)
    maxBufSize : Integer 4096
    aBusSpec : instance of ControlSpec (000002D58E641E48, size=9, set=4)
    cBusSpec : instance of ControlSpec (000002D58E641B08, size=9, set=4)
    cycleSpec : instance of ControlSpec (000002D58E6F83E8, size=9, set=4)
    yZoomSpec : instance of ControlSpec (000002D58E641568, size=9, set=4)
    smallSize : instance of Size (000002D58E739B08, size=2, set=2)
    largeSize : instance of Size (000002D58E739838, size=2, set=2)
    bus : instance of Bus (000002D58DF59AC8, size=5, set=3)
    busSpec : instance of ControlSpec (000002D58E641E48, size=9, set=4)
    cycle : Float 1024.000000   00000000 40900000
    yZoom : Float 1.000000   00000000 3FF00000
    sizeToggle : false
}
ARGS:
Instance of Rect {    (000002D58E873B08, gc=A4, fmt=00, flg=00, set=02)
  instance variables [4]
    left : Integer 200
    top : Integer 200
    width : Integer 400
    height : Integer 400
}
CALL STACK:
	DoesNotUnderstandError:reportError
		arg this = <instance of DoesNotUnderstandError>
	Nil:handleError
		arg this = nil
		arg error = <instance of DoesNotUnderstandError>
	Thread:handleError
		arg this = <instance of Thread>
		arg error = <instance of DoesNotUnderstandError>
	Object:throw
		arg this = <instance of DoesNotUnderstandError>
	Object:doesNotUnderstand
		arg this = <instance of Stethoscope>
		arg selector = 'bounds_'
		arg args = [*1]
	Interpreter:interpretPrintCmdLine
		arg this = <instance of Interpreter>
		var res = nil
		var func = <instance of Function>
		var code = "s.scope.bounds_(Rect(200,200..."
		var doc = nil
		var ideClass = <instance of Meta_ScIDE>
	Process:interpretPrintCmdLine
		arg this = <instance of Main>
^^ ERROR: Message 'bounds_' not understood.
Perhaps you misspelled 'bus_', or meant to call 'bounds_' on another receiver?
RECEIVER: a Stethoscope
```
The full error message for  `s.meter.bounds_(Rect(200,200,400,400))`:
```
ERROR: Message 'bounds_' not understood.
RECEIVER:
Instance of Server {    (000002D58ADAB238, gc=A8, fmt=00, flg=00, set=05)
  instance variables [31]
    name : Symbol 'localhost'
    addr : instance of NetAddr (000002D58ADA6498, size=4, set=2)
    clientID : Integer 0
    isLocal : true
    inProcess : false
    sendQuit : nil
    remoteControlled : false
    maxNumClients : nil
    options : instance of ServerOptions (000002D58ADA9518, size=40, set=6)
    latency : Float 0.200000   9999999A 3FC99999
    dumpMode : Integer 0
    nodeAllocator : instance of NodeIDAllocator (000002D58ADA3898, size=7, set=3)
    controlBusAllocator : instance of ContiguousBlockAllocator (000002D58ADA5AA8, size=6, set=3)
    audioBusAllocator : instance of ContiguousBlockAllocator (000002D58ADA1B68, size=6, set=3)
    bufferAllocator : instance of ContiguousBlockAllocator (000002D58ADA5868, size=6, set=3)
    scopeBufferAllocator : instance of StackNumberAllocator (000002D58ADA6268, size=4, set=2)
    tree : nil
    defaultGroup : instance of Group (000002D58ADA2D78, size=5, set=3)
    defaultGroups : instance of Array (000002D58ADA5CB8, size=1, set=2)
    syncThread : nil
    syncTasks : nil
    window : nil
    scopeWindow : instance of Stethoscope (000002D58E240A18, size=32, set=5)
    serverMeter : instance of ServerMeter (000002D58E82C378, size=2, set=2)
    emacsbuf : nil
    volume : instance of Volume (000002D58ADABE58, size=15, set=4)
    recorder : instance of Recorder (000002D58ADAC418, size=14, set=4)
    statusWatcher : instance of ServerStatusWatcher (000002D58ADAB468, size=20, set=5)
    pid : nil
    serverInterface : nil
    pidReleaseCondition : instance of Condition (000002D58ADA5D28, size=2, set=2)
}
ARGS:
Instance of Rect {    (000002D58E8BCF08, gc=A4, fmt=00, flg=00, set=02)
  instance variables [4]
    left : Integer 200
    top : Integer 200
    width : Integer 400
    height : Integer 400
}
CALL STACK:
	DoesNotUnderstandError:reportError
		arg this = <instance of DoesNotUnderstandError>
	Nil:handleError
		arg this = nil
		arg error = <instance of DoesNotUnderstandError>
	Thread:handleError
		arg this = <instance of Thread>
		arg error = <instance of DoesNotUnderstandError>
	Object:throw
		arg this = <instance of DoesNotUnderstandError>
	Object:doesNotUnderstand
		arg this = <instance of Server>
		arg selector = 'bounds_'
		arg args = [*1]
	Interpreter:interpretPrintCmdLine
		arg this = <instance of Interpreter>
		var res = nil
		var func = <instance of Function>
		var code = "s.meter.bounds_(Rect(200,200..."
		var doc = nil
		var ideClass = <instance of Meta_ScIDE>
	Process:interpretPrintCmdLine
		arg this = <instance of Main>
^^ ERROR: Message 'bounds_' not understood.
RECEIVER: localhost
```

**Comment by prko** [comment]
>or similar (can't test atm) then it may not be worth adding?

While testing this, I found that the following works:
`s.scope.window.bounds_(Rect(100, 100, 400, 400))`
`s.scope.view.bounds_(Rect(100, 100, 400, 400))`

My modification allows the following:
`s.scope(bounds: Rect(100, 100, 400, 400))`

@mtmccrea @telephone 
I think my modification is worth adding because it is shorter. However, if you think that adding my modification to `.scope` is better, I can remove all `.scope` related changes and leave the modification of `.plotTree` and `.method`. 

I am trying to fix the duplicated part of _server-scope.sc_ and the behaviour of changing the size of s.scope by pressing _m_. These will be commented on in the relevant revised sections. I am posting this first because if you think my modification to `.scope` is not necessary, I do not need to revise those code blocks.

**Comment by prko** [comment]
The behaviour of changing scope size is modified. You can see how it works:
https://www.dropbox.com/scl/fi/mpexks8pbaifmukwmmkjc/scope_changing_size_behaviour.mov?rlkey=k1m6ioihy3clo65w09798zv0u&dl=0
Also, the corresponding documentation is updated:
[Update Stethoscope.schelp](https://github.com/supercollider/supercollider/pull/6230/commits/e73d8edbbaa8fc08716faaa596b5482238312241)

File: SCClassLibrary/Common/GUI/tools/Stethoscope.sc
Comment (prko) [reply SCClassLibrary/Common/GUI/tools/Stethoscope.sc:404]: Thanks, done:
https://github.com/supercollider/supercollider/pull/6230#issuecomment-1987043337

**Comment by prko** [comment]
> While testing this, I found that the following works:
> `s.scope.window.bounds_(Rect(100, 100, 400, 400))`
> `s.scope.view.bounds_(Rect(100, 100, 400, 400))`

I used it in my implementation of .scope. The .scope method class of the Server has now been simplified.

File: SCClassLibrary/Common/GUI/PlusGUI/Control/server-scope.sc
Comment (prko) [reply SCClassLibrary/Common/GUI/PlusGUI/Control/server-scope.sc]: Thanks, done!
https://github.com/supercollider/supercollider/pull/6230#issuecomment-1987051825

**Comment by prko** [comment]
>or similar (can't test atm) then it may not be worth adding?

Now there are three options to change the boundary of scope window:
- `s.scope.window.bounds_(Rect(100, 100, 400, 400))` 
- `s.scope.view.bounds_(Rect(100, 100, 400, 400))`
- `s.scope(bounds: Rect(100, 100, 400, 400))` (available if my modification is approved)

@mtmccrea @telephone 
I think my modification is worth adding because it is shorter than the other ways. 

However, if you think that adding my modification to `.scope` is not necessary, I will remove all `.scope` related changes and just leave the modification to `.plotTree` and `.method`.

File: SCClassLibrary/Common/GUI/PlusGUI/Control/server-scope.sc
Comment (telephon) [inline SCClassLibrary/Common/GUI/PlusGUI/Control/server-scope.sc:27]: no need for an explicit `nil` default value. `zoom` is also nil.

Comment (prko) [reply SCClassLibrary/Common/GUI/PlusGUI/Control/server-scope.sc]: Yes! I removed `= nil` in argument in [server-scope.sc](https://github.com/supercollider/supercollider/pull/6230/commits/25d259ab8d27686a0b478ccc8c28390337e84965) and [server-meter.sc](https://github.com/supercollider/supercollider/pull/6230/commits/e76ab825471db0e5b0d0227332ce7045d901611f)

**Comment by mtmccrea** [comment]
Thanks for your patience on this @prko.

>I think my modification is worth adding because it is shorter than the other ways.

My hesitation about adding these as arguments to the creation methods is that I could image more features being added to these UIs in the future, which may require arguments more closely related to their functionality, as opposed to generic view parameters, which can be accessed through the the methods you found above. 

I think a good middle ground may be to add a new `-bounds` instance method to `Stethoscope`. In fact it already has a `-size` method which takes a single argument for it's (square) dimension. (OT: This should probably be extended to take `Size` or a second arg that allows setting a width and/or height, but anyway...)

I did a quick implementation of a `Stethescope:-bounds` method:
```supercollider
	bounds_ { arg rect;
		if( window.notNil ) { window.bounds_(rect.asRect) };
	}
```

So to add to the list of calls you found:
```supercollider
s.scope.window.bounds_(Rect(100, 100, 400, 400)) // current
s.scope.view.bounds_(Rect(100, 100, 400, 400)) // current
s.scope(bounds: Rect(100, 100, 400, 400)) // proposed in this PR
s.scope.bounds_(Rect(100, 100, 400, 400)) // suggested
```

What do you think?

**Comment by mtmccrea** [comment]
Ah, I didn't see that @telephon had started a review... I'm happy to defer to the review on whether my suggestion should be considered. I don't mean to hold up a review that's underway.

**Comment by prko** [comment]
> So to add to the list of calls you found:
> 
> ```supercollider
> s.scope.window.bounds_(Rect(100, 100, 400, 400)) // current
> s.scope.view.bounds_(Rect(100, 100, 400, 400)) // current
> s.scope(bounds: Rect(100, 100, 400, 400)) // proposed in this PR
> s.scope.bounds_(Rect(100, 100, 400, 400)) // suggested
> ```
> 
> What do you think?

Wow! For me, the way you suggested is better to read: 
> s.scope.bounds_(Rect(100, 100, 400, 400)) // suggested

I think @mtmccrea's suggestion could also be applied to `.plotTree' and maybe even `.meter'. What do you think?

I will wait for @telephon's reply.

@mtmccrea and @telephon. Please tell me what to do:
- Remove `s.scope(bounds: Rect(100, 100, 400, 400))` and add @mtmccrea's suggestion.
- Remove all my suggestions in this PR and add @mtmccrea's suggestion to `s.scope`, `s.plotTree` and `s.meter`. I may need help with `s.meter` as the dimension of `s.meter` is automatically calculated by the number of in and out channels.
- Just remove `s.scope(bounds: Rect(100, 100, 400, 400))`. @mtmccrea adds @mtmccrea's suggestion.
- Close this PR. @mtmccrea adds @mtmccrea's suggestion to `s.cope', `s.plotTree` and `s.meter`.

**Comment by mtmccrea** [comment]
> Remove all my suggestions in this PR and add @mtmccrea's suggestion to s.scope, s.plotTree and s.meter. I may need help with s.meter as the dimension of s.meter is automatically calculated by the number of in and out channels.

With a quick look at all these methods, I see that their internal handling of the created Window is different for between them, so the solution won't be the same for each. It's good to have a consistent approach to setting bounds across all server-related GUIs, but this may not be possible. 

In general, we try to avoid adding new arguments if the solution can be achieved through method calls that are just as concise.

The `.scope` solution I mentioned is fairly straightforward, so this PR could be confined to just that if the other methods require further research/experimentation on your part.

**Comment by prko** [comment]
@mtmccrea 
I will remove
- the argument `bounds` in `.scope` and `.plotTree`;
- the argument `position` in `.meter` in this PR.

Then I will add the `-.bounds_` method to the `Stethescope` class as you suggested.

Then I will open two new PRs:
- One PR to add the `-.bounds_` method to the `Server` class in the `ServerPlusGUI` file.
   The `serverTree` in the current PR allows only one instance of `serverTree`, and I think it is useful.
- A PR to add the `-.pos_` method to the `Server` class in the `server-meter.sc` file. 
  .position_` is too long, so I want to use `.pos_`.
However, to create two new PRs, `.meter` and `.plotTree` should return a GUI element, but currently each returns nothing. As far as I have checked using `Find` in VScode, there are no code examples that follow other methods after `.meter` and `.plotTree`. So I think this is not a problem.

What do you think of this idea?

**Comment by prko** [comment]
I removed almost all changes except for this scope window resize behaviour, and added `.bounds_` to `Stethoscope`.

**Comment by prko** [comment]
Oh, sorry!  Because I changed the name of the connected branch to this PR, the PR is closed and I can connect this PR and the renamed one. I will open a new PR. Sorry, I am still a beginner in this kind of work....

**Comment by prko** [comment]
@mtmccrea I think I have done everything you mentioned. I just hope my activation is helpful for the development of SC.

**Pull request closed without merging**
